Introducing probes and sending raw timers to ServiceControl

tmasternak opened the pull request:
Supersedes #32
Connects to https://github.com/Particular/LaunchWave.DevOpsAdvancedMonitoring/issues/52

## Description
This PR and #33 introduces changes that are result of requirements of https://github.com/Particular/LaunchWave.DevOpsAdvancedMonitoring/issues/52. The number one needed is to be able to ship raw `Processing Time` and `Critical Time` measurements to `ServiceControl.Monitoring`. The reason this is needed is that we need ability to calculate percentiles over those measurements and aggregate them with different time-windows (minutes, hours, days).

To enable raw data capturing a concept of `Probe` has been introduced - it represents a source of measurements. With that approach `Processing Time` and `Critical Time` are `DurationProbes` that are observed and measurements are shipped to `ServiceControl.Monitoring`.

The public API has been changed as well. Currently there are two methods. First which enables sending raw data to `SericeControl.Monitoring` and second which enables attaching observers to `Probes`. In removes dependency on Metrics.Net which has pretty opinionated approach to aggregating timing data.

#32 mentioned was a previous attempt to mitigate some of the problems mentioned which finally turned out to be a dead alley. It did not enable calculating averages in a flexible enough manner to work with period filter in SP wireframes and did not provide ability to calculate percentiles.

## Scope
This PR is meant to make following changes:
  * [x] Introduce concept of `Probe` that represents source generating measurement data that can be observed
  * [x] Migrate `ProcessingTime`, `CriticalTime` and performance meters to Probes
  * [x] Removes all report types except for `ServiceControl` report - 3rd party integration will happen at the level of Probes
  * [x] Removes `ProcessingTime` and `CriticalTime` from Metrics.Net based report into separate message
    * [x] Remove timing measurements from Metrics.Net report
    * [x] Introduce separate message type is provided via #33
  * [x] Validate that this approach will work with `NServiceBus.Metrics.PerformanceCounters`
    * [x] Removes MetricsAttributes which will require changes to `NServiceBus.Metrics.PerformanceCounters` to scan for `ProbeProperties` attribute instead.
   * [x] Create `WIP` [PR](https://github.com/Particular/NServiceBus.Metrics.PerformanceCounters/pull/19)

Scooletz commented on src/NServiceBus.Metrics/MetricsOptions.cs:
```diff
@@ -19,69 +11,131 @@ public class MetricsOptions
-        /// <param name="interval">How often metric data is sent to the log</param>
-        /// <param name="logLevel">Level at which log entries should be written. Default is DEBUG.</param>
-        public void EnableLogTracing(TimeSpan interval, LogLevel logLevel = LogLevel.Debug)
+        public SignalProbe[] Signals;
```
The default ctor can be made internal

tmasternak replied:
I've also changes properties to make sure that those cannot be changed when registering observers.

Scooletz commented on src/NServiceBus.Metrics/Probes/DurationProbeBuilder.cs:
```diff
@@ -0,0 +1,42 @@
+
+        DurationProbe GetProbe()
+        {
+            var attr = GetType().GetCustomAttribute<ProbePropertiesAttribute>();
```
This requires a null check, and throwing a user friendly exception like below, if not attribute found.

`$The type '{GetType()}' is not annotated with 'ProbePropertiesAttribute'. Add this attribute for providing probe's metadata`

tmasternak replied:
Adjusted the text a bit:

> $"The type '{GetType()}' is not annotated with required '{typeof(ProbePropertiesAttribute).Name}'. This attribute has to be added to provide necessary metadata for the probe."

Scooletz commented on src/NServiceBus.Metrics/Probes/DurationProbeBuilder.cs:
```diff
@@ -0,0 +1,42 @@
+        }
+    }
+
+    class ProbePropertiesAttribute : Attribute
```
`AttributeUsage` -> class only, singular only
seal the attribute

tmasternak replied:
Added. Single usage is by default - I did not make it explicit.

Scooletz commented on src/NServiceBus.Metrics/Probes/SignalProbeBuilder.cs:
```diff
@@ -0,0 +1,27 @@
+
+    SignalProbe GetProbe()
+    {
+        var attr = GetType().GetCustomAttribute<ProbePropertiesAttribute>();
```
This requires a null check, and throwing a user friendly exception like below, if not attribute found.

$The type '{GetType()}' is not annotated with 'ProbePropertiesAttribute'. Add this attribute for providing probe's metadata

tmasternak commented:
I've added two changes to this PR during migration of `NServiceBus.Metrics.PerformanceCounters`:
 * Changes needed to enable testing: public ctor on `ProbesContext` and `ISignalProbe` and `IDurationProbe`
 * Reverted back names and descriptions of probes to be the same as for meters and timers - this is needed for backwards compatibility perspective but probably should be handled in `NServiceBus.Metrics.PerformanceCounters`

/cc: @Scooletz

Scooletz commented:
@tmasternak This is a minor, but now this PR has `ISignalProbe` and `IDurationProbe` in its public API, which, both, provide `Name` & `Description` from `Probe` base class. Should we provide `IProbe` in public API to gather common properties, implemented by `Probe` base class?

Scooletz commented:
@mikeminutillo @dvdstelt @WilliamBZA Could you chime in with your reviews?

Scooletz commented:
LATM (Looks Awesome To Me) @tmasternak

Scooletz commented:
I think that @tmasternak proposal, with simple delegate `+=` covers our needs. To address being reactive and able to batch process multiple signals in a well-performing way we've got #33

Scooletz commented:
Ping @Particular/metrics-maintainers

danielmarbach commented on src/NServiceBus.Metrics.Tests/NServiceBus.Metrics.approved.cs:
```diff
@@ -13,10 +26,14 @@ public class static MetricsConfigurationExtensions
-        public void EnableCustomReport(System.Func<string, System.Threading.Tasks.Task> func, System.TimeSpan interval) { }
-        public void EnableLogTracing(System.TimeSpan interval, NServiceBus.Logging.LogLevel logLevel = 0) { }
-        public void EnableMetricTracing(System.TimeSpan interval) { }
+        public void RegisterObservers(System.Action<NServiceBus.ProbeContext> register) { }
```
See https://github.com/Particular/NServiceBus.Metrics.PerformanceCounters/pull/19#discussion_r124504286

tmasternak replied:
I consider this fine after reading the conclusion in the comment mentioned.

danielmarbach commented on src/NServiceBus.Metrics/MetricsFeature.cs:
```diff
@@ -1,63 +1,102 @@
+    static void SetUpSignalReporting(ProbeContext probeContext, MetricsContext metricsContext)
     {
-        var builders = context.Settings.Get<MetricsRegistry>().Builders;
+        probeContext.Signals.ToList().ForEach(sp =>
```
Closure access?

danielmarbach replied:
use foreach?

tmasternak replied:
I have chosen `ForEach` for readability. It is not the hot path and executed only during feature configuration.

danielmarbach commented on src/NServiceBus.Metrics/MetricsFeature.cs:
```diff
@@ -1,63 +1,102 @@
+
+    static void SetUpQueueLengthReporting(FeatureConfigurationContext context, MetricsContext metricsContext)
+    {
+        new QueueLengthTracker().SetUp(metricsContext, context);
```
Why not just static method?

danielmarbach commented on src/NServiceBus.Metrics/MetricsOptions.cs:
```diff
@@ -19,69 +11,105 @@ public class MetricsOptions
         /// </summary>
         /// <param name="serviceControlMetricsAddress">The transport address of the ServiceControl instance</param>
-        /// <param name="interval">How frequently metric data is sent to ServiceControl</param>
+        /// <param name="interval">Inteval between consequitive reports</param>
```
Spelling of interval

SimonCropp replied:
consequitive

danielmarbach commented on src/NServiceBus.Metrics/MetricsOptions.cs:
```diff
@@ -19,69 +11,105 @@ public class MetricsOptions
+    /// <summary>
+    /// Stores avialbe probes
+    /// </summary>
+    public class ProbeContext
```
Move out to different files?

danielmarbach commented on src/NServiceBus.Metrics/MetricsOptions.cs:
```diff
@@ -19,69 +11,105 @@ public class MetricsOptions
+        /// Enables registering action called on event occurence.
+        /// </summary>
+        /// <param name="observer"></param>
+        void Register(Action observer);
```
See comment about state and closure access

danielmarbach commented on src/NServiceBus.Metrics/MetricsOptions.cs:
```diff
@@ -19,69 +11,105 @@ public class MetricsOptions
+        /// Enables registering action called on event occurence.
+        /// </summary>
+        /// <param name="observer"></param>
+        void Register(Action<TimeSpan> observer);
```
See comment about state and closure access

danielmarbach commented on src/NServiceBus.Metrics/ProbeBuilders/DurationProbeBuilder.cs:
```diff
@@ -0,0 +1,36 @@
+        return new DurationProbe(attribute.Name, attribute.Description);
+    }
+
+    protected static readonly Task<int> CompletedTask = Task.FromResult(0);
```
Move out to TaskExtensions?

danielmarbach commented on src/NServiceBus.Metrics/Probes/DurationProbe.cs:
```diff
@@ -0,0 +1,21 @@
+
+    public void Register(Action<TimeSpan> observer)
+    {
+        observers += observer;
```
Not sure it is worth discussing but isn't this keeping a strong reference to the observer? Could this leak?

Scooletz replied:
My assumption that you register to probes forever. I see no value for making them unregisterable/weak. Do you have an example for it @danielmarbach  ?

danielmarbach replied:
Starting and stopping endpoints? Dynamic endpoints?

Scooletz replied:
But the probe is attached to the endpoint. It's done in `MetricsFeature`. When it dies, it takes its probes with itself. I see that your examples are re-entrant in terms of, that the same process can spawn multiple endpoints of the same type. Still, I see no flaw i it.

danielmarbach commented on src/NServiceBus.Metrics/Probes/SignalProbe.cs:
```diff
@@ -0,0 +1,21 @@
+
+    public void Register(Action observer)
+    {
+        observers += observer;
```
See leak question above

danielmarbach commented:
Yeah correct since the delegates are bound to the probe instance we are fine

Szymon Kulec <notifications@github.com> schrieb am Mi. 28. Juni 2017 um
16:54:

> *@Scooletz* commented on this pull request.
> ------------------------------
>
> In src/NServiceBus.Metrics/Probes/DurationProbe.cs
> <https://github.com/Particular/NServiceBus.Metrics/pull/35#discussion_r124562599>
> :
>
> > @@ -0,0 +1,21 @@
> +﻿using System;
> +using NServiceBus;
> +
> +class DurationProbe : Probe, IDurationProbe
> +{
> +    public DurationProbe(string name, string description) : base(name, description)
> +    {
> +    }
> +
> +    public void Register(Action<TimeSpan> observer)
> +    {
> +        observers += observer;
>
> But the probe is attached to the endpoint. It's done in MetricsFeature.
> When it dies, it takes its probes with itself. I see that your examples are
> re-entrant in terms of, that the same process can spawn multiple endpoints
> of the same type. Still, I see no flaw i it.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Particular/NServiceBus.Metrics/pull/35#discussion_r124562599>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAKostx5iqp0mLbXtZyuJO7dNnerHO4-ks5sImg3gaJpZM4OEC_T>
> .
>
-- 

Have you tried NServiceBus V6 <https://particular.net/nservicebus-v6> yet?

Learn how to scale applications with microservices and NServiceBus 6
<https://www.pluralsight.com/courses/microservices-nservicebus6-scaling-applications>
.
------------------------------

*Join Udi Dahan's next Advanced Distributed Design Course:*

September 2017 Los Angeles, CA, USA <http://particular.net/adsd>

November 2017 Sydney, Australia <http://particular.net/adsd>
------------------------------

SimonCropp commented on src/NServiceBus.Metrics/MetricsOptions.cs:
```diff
@@ -19,69 +11,105 @@ public class MetricsOptions
+    }
+
+    /// <summary>
+    /// Stores avialbe probes
```
Typo

SimonCropp commented on src/NServiceBus.Metrics/MetricsOptions.cs:
```diff
@@ -19,69 +11,105 @@ public class MetricsOptions
+    {
         /// <summary>
-        /// Enables sending metric data to the NServiceBus log
+        /// Crates <see cref="ProbeContext"/>.
```
Typo

SimonCropp commented on src/NServiceBus.Metrics/MetricsOptions.cs:
```diff
@@ -19,69 +11,105 @@ public class MetricsOptions
-        /// <param name="logLevel">Level at which log entries should be written. Default is DEBUG.</param>
-        public void EnableLogTracing(TimeSpan interval, LogLevel logLevel = LogLevel.Debug)
+        /// <param name="durations">Duration probes collection.</param>
+        /// <param name="signals">Signal probes collection</param>
```
These two param comments dont really add any value. Suggest either fleshing out the description or removing them

SimonCropp commented on src/NServiceBus.Metrics/MetricsOptions.cs:
```diff
@@ -19,69 +11,105 @@ public class MetricsOptions
+        /// <summary>
+        /// Enables registering action called on event occurence.
+        /// </summary>
+        /// <param name="observer"></param>
```
Remove the redundant doco node

SimonCropp commented on src/NServiceBus.Metrics/MetricsOptions.cs:
```diff
@@ -19,69 +11,105 @@ public class MetricsOptions
+        /// <summary>
+        /// Signal type probes.
+        /// </summary>
+        public ISignalProbe[] Signals { get; }
```
So this is a public api? Is it ok for people to manipulate  the array?

tmasternak replied:
Good point. Switching to `ReadonlyCollection`

SimonCropp commented on src/NServiceBus.Metrics/MetricsOptions.cs:
```diff
@@ -19,69 +11,105 @@ public class MetricsOptions
+    /// <summary>
+    /// Probe that measures duration of an event.
+    /// </summary>
+    public interface IDurationProbe : IProbe
```
Can these types be split to their own files?

SimonCropp commented on src/NServiceBus.Metrics.AcceptanceTests/Tests/QueueLength/When_sending_message.cs:
```diff
@@ -38,26 +37,43 @@ public class When_sending_message : NServiceBusAcceptanceTest
+            }
+            catch
+            {
+                return false;
```
Why not just let it bubble?

tmasternak replied:
Because we want to check the asserts until they are true or test execution time elapses. If we would throw here the test would end in failure potentially prematurely.

The general problem here is that we don't have a reasonable way of saying when is the time report should include sequence data.

SimonCropp commented on src/NServiceBus.Metrics/MetricsFeature.cs:
```diff
@@ -1,63 +1,102 @@
+        SetUpSignalReporting(probeContext, metricsContext);
 
-        context.RegisterStartupTask(builder => new MetricsReporting(metricsContext, metricsOptions, builder));
+        if (!string.IsNullOrEmpty(metricsOptions.ServiceControlMetricsAddress))
```
Shouldnt empty string be an error case

WilliamBZA replied:
It's possible to use the metrics without a metrics server, metrics server is only for sending the data to e.g. SP.

SimonCropp replied:
so in that case wouldnt it be null?

tmasternak replied:
No in that case the method would not be called and only  `RegisterObservers` would be used.

SimonCropp commented on src/NServiceBus.Metrics/ProbeBuilders/SignalProbeBuilder.cs:
```diff
@@ -0,0 +1,30 @@
+
+        if (attribute == null)
+        {
+            var exceptionMessage = $"The type '{GetType()}' is not annotated with required '{typeof(ProbePropertiesAttribute).Name}'. This attribute has to be added to provide necessary metadata for the probe.";
```
Seems very similar to the above attribute interrogation. Perhaps make them an extension method?

tmasternak replied:
> Seems very similar to the above attribute interrogation.

I'm not sure I'm following.

dvdstelt replied:
Me neither. Both `DurationProbeBuilder` and `SignalProbeBuilder` look 100% alike, if I'm not mistaken. Perhaps you can create a generic class to eliminate a class?

dvdstelt commented:
Added "Connects to https://github.com/Particular/LaunchWave.DevOpsAdvancedMonitoring/issues/52" to description.

danielmarbach commented on src/NServiceBus.Metrics/TaskExtensions.cs:
```diff
@@ -0,0 +1,9 @@
+
+    class TaskExtensions
+    {
+        public static Task Completed => Task.FromResult(0);
```
Make the class static and assign the static field once. See https://github.com/ahofman/NServiceBus/blob/develop/src/NServiceBus.Core/TaskEx.cs#L19

kbaley commented on src/NServiceBus.Metrics/MetricsOptions.cs:
```diff
@@ -19,69 +11,36 @@ public class MetricsOptions
 
         /// <summary>
-        /// Enables sending metric data to the trace log
+        /// Enables registering observers to available probes.
```
Doesn't really _enable_ the registering so much as it actually _does_ register observers.

tmasternak replied:
I wouldn't say that it registers observers. What I meant by `enables` is that `Action<ProbeContext>` passed to to `RegisterObservers` can call e.g. 
```
probeContext.Signals.First().Register(() => ... )
```

kbaley commented on src/NServiceBus.Metrics/ProbeBuilders/DurationProbeBuilder.cs:
```diff
@@ -0,0 +1,33 @@
+
+        if (attribute == null)
+        {
+            var exceptionMessage = $"The type '{GetType()}' is not annotated with required '{typeof(ProbePropertiesAttribute).Name}'. This attribute has to be added to provide necessary metadata for the probe.";
```
Could tighten this up:

> The type '{GetType()}' is not annotated with required '{typeof(ProbePropertiesAttribute).Name}'. Add this attribute to provide necessary metadata for the probe.

tmasternak commented on src/NServiceBus.Metrics.AcceptanceTests/Tests/QueueLength/When_publishing_message.cs:
```diff
@@ -38,26 +37,45 @@ public class When_publishing_message : NServiceBusAcceptanceTest
-            var data = JObject.Parse(context.Data);
-            var counters = (JArray)data["Counters"];
-            var counterTokens = counters.Where(c => c.Value<string>("Name").StartsWith("Sent sequence for"));
+            SequencesReported(context);
```
> This call is included in the .Done condition above so doesn't that mean that all the checks have been done by this point?

Removed the unneeded call at the end of the test.

> Plus, all the asserts in it are wrapped in a try-catch so this call doesn't really do anything.

Try-catch is needed to return true/false from the method instead of throwing. The reason why this condition has to be checked in `.Done` is that there is no reliable way to know when the metrics report should contain sequence data unless the reporting is broken. Is is caused by the fact that report is sent async to processing messages. In effect this test will wait until sequences are reported or time-out and fail.

tmasternak commented:
@danielmarbach can you have a look?

danielmarbach commented on src/NServiceBus.Metrics.AcceptanceTests/Tests/QueueLength/When_publishing_message.cs:
```diff
@@ -38,26 +37,43 @@ public class When_publishing_message : NServiceBusAcceptanceTest
+        }
 
-            var sessionIds = new[] { AssertHeaders(context.Headers1), AssertHeaders(context.Headers2) };
+        static bool SequencesReported(Context context)
```
I do not think this is a good idea. SequencesReported will be executed many times by the done condition infrastructure and it contains business asserts that should be part of the test not burried in the done condition checking. I suggest to have a simple done condition and then do all the parsing and asserts after the await of the scenario in a dedicated Assert method.

tmasternak replied:
Can you have a look at https://github.com/Particular/NServiceBus.Metrics/pull/35#discussion_r125588411?

danielmarbach commented on src/NServiceBus.Metrics.AcceptanceTests/Tests/QueueLength/When_sending_message.cs:
```diff
@@ -38,26 +37,43 @@ public class When_sending_message : NServiceBusAcceptanceTest
+        }
 
-            var sessionIds = new [] { AssertHeaders(context.Headers1), AssertHeaders(context.Headers2)};
+        static bool SequencesReported(Context context)
```
Same comment as above

danielmarbach commented on src/NServiceBus.Metrics.Tests/NServiceBus.Metrics.approved.cs:
```diff
@@ -13,10 +26,24 @@ public class static MetricsConfigurationExtensions
     public class MetricsOptions
     {
         public MetricsOptions() { }
+        [System.ObsoleteAttribute("Use RegisterObservers instead to attach to monitoring probes. Will be treated as " +
```
How about

`Use RegisterObservers instead to attach to probes` ?

danielmarbach commented on src/NServiceBus.Metrics.Tests/NServiceBus.Metrics.approved.cs:
```diff
@@ -5,6 +5,19 @@
+    }
+    public interface ISignalProbe : NServiceBus.IProbe
+    {
+        void Register(System.Action observer);
```
If this is called Signal probe what is the signal then since action takes no input

tmasternak replied:
Calling of the action is the signal.

danielmarbach commented on src/NServiceBus.Metrics/IDurationProbe.cs:
```diff
@@ -0,0 +1,17 @@
+        /// <summary>
+        /// Enables registering action called on event occurrence.
+        /// </summary>
+        /// <param name="observer"></param>
```
Should the param be filled out with a meaningful description?

tmasternak replied:
removed as in `ISingalProbe`

danielmarbach commented on src/NServiceBus.Metrics/IProbe.cs:
```diff
@@ -0,0 +1,18 @@
+﻿namespace NServiceBus
+{
+    /// <summary>
+    /// Represents probe description.
```
`Represent the probe description` ?

tmasternak replied:
Changed to `Provides general information about a probe`

danielmarbach commented on src/NServiceBus.Metrics/IProbe.cs:
```diff
@@ -0,0 +1,18 @@
+        string Name { get; }
+
+        /// <summary>
+        /// Descripton of the probe.
```
Spelling in Descripton

danielmarbach commented on src/NServiceBus.Metrics/ISignalProbe.cs:
```diff
@@ -0,0 +1,15 @@
+        /// <summary>
+        /// Enables registering action called on event occurrence.
+        /// </summary>
+        void Register(Action observer);
```
I think with this description it is a bit clearer. It is just an event that occurred without any payload or meaningful measurement. The event name is the signal right?

tmasternak replied:
Correct.

danielmarbach commented on src/NServiceBus.Metrics/ProbeBuilders/DurationProbeBuilder.cs:
```diff
@@ -0,0 +1,33 @@
+using System.Reflection;
+
+/// <summary>
+/// Defines a custom metric.
```
Remove comment?

danielmarbach commented on src/NServiceBus.Metrics/ProbeBuilders/ReceivePerformanceDiagnosticsBehavior.cs:
```diff
@@ -0,0 +1,27 @@
+using System.Threading.Tasks;
+using NServiceBus.Pipeline;
+
+class ReceivePerformanceDiagnosticsBehavior : IBehavior<IIncomingPhysicalMessageContext, IIncomingPhysicalMessageContext>
```
The core has also a behavior with the same name. Did we make sure that it does not collapse?

tmasternak replied:
Hmm the name of the behavior has [not changed](https://github.com/Particular/NServiceBus.Metrics/blob/develop/src/NServiceBus.Metrics/Performance/ReceivePerformanceDiagnosticsBehavior.cs) with this PR so I assume it's fine.

dvdstelt commented:
Hmmm, when adding comments in a review and then selecting to add the review, but instead of approving/request-changes add review as comments, the request for review stays the same?
I added a comment so I'll just approve it from here on.

tmasternak commented:
@danielmarbach done :)
